src/tpm2_pytss/constants.py

```diff
@@ -200,12 +200,28 @@ def __xor__(self, value):
         return self.__class__(int(self).__xor__(value))
 
     @staticmethod
-    def _fix_const_type(cls):
-        for k, v in vars(cls).items():
+    def _copy_and_set(dstcls, srccls):
+        """Copy class variables from srccls to dstcls
+
+        Sets the the variable type to dstcls when copying.
+        dstcls and srccls can be the same.
+        """
+        for k, v in vars(srccls).items():
             if not isinstance(v, int) or k.startswith("_"):
                 continue
-            fv = cls(v)
-            setattr(cls, k, fv)
+            fv = dstcls(v)
+            setattr(dstcls, k, fv)
+
+    @staticmethod
+    def _fix_const_type(cls):
+        TPM_FRIENDLY_INT._copy_and_set(cls, cls)
+        # check for superclasses which are subclasses of TPM_FRIENDLY_INT, but isn't
+        # a TPM_FRIENDLY_INT or TPMA_FRIENDLY_INTLIST and copy the constants from the
+        # parent class. This enables good inheritance for the different constants types.
+        for sc in cls.__mro__[1:]:
```

Comment:
doesn't this work? We essentially don't care about object, int, and the superclass helper which are the top 3 in the inheritance list (see output of mro slice below). I would like to avoid hardcoding class names, but I guess relying on inheritance structure could be pretty brittle too. I guess its two versions of the same problem. Your thoughts?

```python
for sc in cls.__mro__[1:-3]:
```

```python3
>>> TPM2_ECC.__mro__[1:-3]
()
>>> TPM2_ALG_ID.__mro__[1:-3]
(<class 'tpm2_pytss.constants.TPM2_ALG'>,)
```

Reply:
@whooo this patch ontop of yours works for me:
```git
diff --git a/src/tpm2_pytss/constants.py b/src/tpm2_pytss/constants.py
index a2acd127e321..520f22b52b12 100644
--- a/src/tpm2_pytss/constants.py
+++ b/src/tpm2_pytss/constants.py
@@ -214,13 +214,18 @@ class TPM_FRIENDLY_INT(int):
 
     @staticmethod
     def _fix_const_type(cls):
-        TPM_FRIENDLY_INT._copy_and_set(cls, cls)
-        # check for superclasses which are subclasses of TPM_FRIENDLY_INT, but isn't
-        # a TPM_FRIENDLY_INT or TPMA_FRIENDLY_INTLIST and copy the constants from the
-        # parent class. This enables good inheritance for the different constants types.
-        for sc in cls.__mro__[1:]:
-            if sc in (TPM_FRIENDLY_INT, TPMA_FRIENDLY_INTLIST):
-                break
+        # We want to copy class variables from the base class and any "intermediate" classes.
+        #
+        # Skip the:
+        # - the root "object" class
+        # - Helper classes TPM_FRIENDLY_INT or TPM_FRIENDLY_INTLIST
+        # consider the following examples showing inheritance with
+        # a helper class from the list above and without.
+        # >>> TPM2_ECC.__mro__[:-3]
+        # (<class 'tpm2_pytss.constants.TPM2_ECC'>,)
+        # >>> TPM2_ALG_ID.__mro__[:-3]
+        # (<class 'tpm2_pytss.constants.TPM2_ALG_ID'>, <class 'tpm2_pytss.constants.TPM2_ALG'>)
+        for sc in cls.__mro__[:-3]:
             TPM_FRIENDLY_INT._copy_and_set(cls, sc)
         return cls
 ```

Reply:
My worry is that it would create surprises in the future if it doesn't follow a strict inheritance scheme, lets say someone wants to add a mixin class and writes:
```python3
@TPM_FRIENDLY_INT._fix_const_type
TPM2_COOL_CONSTANT(MyMixin, TPM2_ECC):
    pass
```
That would break that flow and it wouldn't be obvious why without digging a bit in the code.
Looking at my current code it would still copy some things from the mixin, which it shouldn't.
So what about, find the first superclass that is a subclass of TPM_FRIENDLY_INT but isn't TPM_FRIENDLY_INT or TPMA_FRIENDLY_INTLIST and copy the constants from that?

I'm also looking at making sure that one could override a value in a subclass

Reply:
> My worry is that it would create surprises in the future if it doesn't follow a strict inheritance scheme, lets say someone wants to add a mixin class and writes:
> 
> ```python
> @TPM_FRIENDLY_INT._fix_const_type
> TPM2_COOL_CONSTANT(MyMixin, TPM2_ECC):
>     pass
> ```
> 
> That would break that flow and it wouldn't be obvious why without digging a bit in the code. Looking at my current code it would still copy some things from the mixin, which it shouldn't. So what about, find the first superclass that is a subclass of TPM_FRIENDLY_INT but isn't TPM_FRIENDLY_INT or TPMA_FRIENDLY_INTLIST and copy the constants from that?
> 
> I'm also looking at making sure that one could override a value in a subclass

If fine with keeping it dirt simple as either case we assume a certain inheritance hierarchy and your code is more verbose about that.

```diff
diff --git a/src/tpm2_pytss/constants.py b/src/tpm2_pytss/constants.py
index a2acd127e321..2ea183d271eb 100644
--- a/src/tpm2_pytss/constants.py
+++ b/src/tpm2_pytss/constants.py
@@ -214,11 +214,10 @@ class TPM_FRIENDLY_INT(int):
 
     @staticmethod
     def _fix_const_type(cls):
-        TPM_FRIENDLY_INT._copy_and_set(cls, cls)
         # check for superclasses which are subclasses of TPM_FRIENDLY_INT, but isn't
         # a TPM_FRIENDLY_INT or TPMA_FRIENDLY_INTLIST and copy the constants from the
         # parent class. This enables good inheritance for the different constants types.
-        for sc in cls.__mro__[1:]:
+        for sc in cls.__mro__:
             if sc in (TPM_FRIENDLY_INT, TPMA_FRIENDLY_INTLIST):
                 break
             TPM_FRIENDLY_INT._copy_and_set(cls, sc)
```
What do you think of this simplification, we copy the baseclass as well as the parent classes so long as they are not TPM_FRIENDLY...?

Reply:
Looks better to me, I'll just split out the changes to how `_fix_const_type` works into its own commit